TACTSharp/RootInstance.cs

```diff
@@ -47,264 +47,259 @@ public enum LocaleFlags : uint
-            public Span<byte> AsSpan() => MemoryMarshal.CreateSpan(ref _element, Length);
-        }
+            // Skip the header.
+            fileData = fileData[headerSize..];
 
-        public RootEntry? GetEntryByFDID(uint fileDataID)
-        {
-            if (entriesFDID.TryGetValue(fileDataID, out var entry))
-                return entry;
+            var pages = new List<Page>();
+            while (fileData.Length != 0)
+            {
+                var recordCount = fileData.ReadInt32LE();
+                fileData = fileData[4..];
 
-            return null;
-        }
+                var (contentFlags, localeFlags) = ParseManifestPageFlags(ref fileData, version);
 
-        public RootEntry? GetEntryByLookup(ulong lookup)
-        {
-            if (entriesLookup.TryGetValue(lookup, out var entry))
-                return entry;
+                // No records in this file.
+                if (recordCount == 0)
+                    continue;
 
-            return null;
-        }
+                // Calculate block size
+                var blockSize = 4 * recordCount; // FileDataID[n]
+                blockSize += MD5.Length * recordCount;
+                if (format == Format.Legacy || !contentFlags.HasFlag(ContentFlags.NoNames))
+                    blockSize += 8 * recordCount;
 
-        public uint[] GetAvailableFDIDs()
-        {
-            return [.. entriesFDID.Keys];
+                var blockData = fileData[..blockSize];
+                fileData = fileData[blockSize..];
+
+                // Determine conditions related to keeping this page.
+                var localeSkip = !localeFlags.HasFlag(LocaleFlags.All_WoW) && !localeFlags.HasFlag(Settings.Locale);
+                var contentSkip = contentFlags.HasFlag(ContentFlags.LowViolence);
+
+                if (localeSkip || contentSkip)
+                    continue;
+
+                // Read a FDID delta array from the file (+1 implied) and adjust instantly.
+                var fdids = blockData.ReadUInt32LE(recordCount);
+                for (var i = 1; i < fdids.Length; ++i)
+                    fdids[i] += fdids[i - 1] + 1;
+
+                // Get a span over the actual record block of this page
+                var recordData = blockData[(4 * recordCount)..];
+                var (records, nameHashRange) = format switch
+                {
+                    Format.Legacy => ParseLegacy(recordData, recordCount, fdids),
+                    Format.MFST => ParseManifest(recordData, recordCount, contentFlags, fdids),
+                    _ => throw new UnreachableException()
+                };
+
+                var nameHashes = nameHashRange.AsEnumerator(recordData);
+                var page = new Page(records, contentFlags, localeFlags);
+
+                // TODO: This happens. What do we do?!
+                // System.ArgumentException: An item with the same key has already been added. Key: 11470997404861800962
```

Comment:
I assume for different content flags?

I guess the perfect world would store hashes by content flag/locale. The hacky workaround would probably be "if current page is a locale one, override, otherwise ignore".

Reply:
For now this is the old behavior which was, I believe, keeping only the first entry encountered. I also believe it has to do with content/locale flags combinations.